ckanext/xloader/plugin.py

```diff
@@ -107,9 +107,23 @@ def notify(self, entity, operation):
         See: ckan/model/modification.py.DomainObjectModificationExtension
         """
         if operation != DomainObjectOperation.changed \
-        or not isinstance(entity, Resource) \
-        or not getattr(entity, 'url_changed', False):
+        or not isinstance(entity, Resource):
             return
+
+        # If the resource requires validation, stop here if validation
+        # has not been performed or did not succeed. The Validation
+        # extension will call resource_patch and this method should
+        # be called again. However, url_changed will not be in the entity
+        # once Validation does the patch.
+        if _is_validation_plugin_loaded() and \
+        toolkit.asbool(toolkit.config.get('ckanext.xloader.requires_validation')):
+            if entity.__dict__.get('extras', {}).get('validation_status', None) != 'success':
```

Comment:
What about resources without a schema? Even if the flag is set, they won't have a validation status, I think?

Reply:
same as above

Reply:
@ThrawnCA the default resource schema in core for `__extras` will accept anything. And the `ckanext-validation` plugin will patch resources with `validation_status`. So this should be fine unless if someone has modified the resource schema to not have extras and/or validation_status.

But I feel like if they have done that, then they are probably not using the validation extension.

Reply:
We're running the validation plugin, but resources that don't specify any schema don't get a validation run or a `validation_status`. Eg https://dev.data.qld.gov.au/dataset/testing-qoldev-687/resource/de2f6122-1dee-4295-a589-8dfb702f5983 (https://dev.data.qld.gov.au/api/action/resource_show?id=de2f6122-1dee-4295-a589-8dfb702f5983)

Reply:
That's interesting. In our case we want all resources validated with a static schema (our users can't be trusted to provide one it seems) before being loaded.

Maybe we could support both cases with another config option, or by defining a helper function that determines whether a resource can be loaded that a theme plugin could then override?

Reply:
In our situation. I do remember our business owner not wanting users who upload files to get annoyed if they provided csv's that were not valid even to basic rules and as such is skipped if schema is missing. 

Most annoying they did not do this as a flag and hard coded in our fork many many moons ago.
https://github.com/qld-gov-au/ckanext-validation/commit/0e42ce7dcd39bd6ba3cc17d94cf70a94b2c6a012#diff-9254636ee2e98ff1eaf809e6f490d1d2193f794c434d9e6e5235e3f2ccdfa717

Reply:
Hmmm @ThrawnCA @duttonw 

(Firstly my bad, I was thinking of the core code schemas not the `schema` field from the Validation plugin)

So I am not sure if the two config options would fully support your setup.

Mainly because even if I add in the second config option and check if the Resource's `schema` field is empty. You have this code (https://github.com/qld-gov-au/ckanext-validation/commit/0e42ce7dcd39bd6ba3cc17d94cf70a94b2c6a012#diff-35a1e13aaa4a67e7f831c29f02207acce46ed14129c73d3ecd1537bb9dcc89f9R77):
```
def _get_schema_or_default_schema(resource):
    schema = resource.get('schema')

    if asbool(resource.get('align_default_schema')):
        dataset = get_action('package_show')({}, {'id': resource['package_id']})

        if dataset:
            schema = dataset.get(u'default_data_schema')

    if schema and isinstance(schema, string_types):
        schema = schema if is_url_valid(schema) else json.loads(schema)

    return schema
```

So if a Resource does not have a `schema`, they can use its package's `default_data_schema`?

If that is the case, then if you have the new config option to require a validation report BUT then ignore empty Resource Schemas, it would then bypass this functionality of the package `default_data_schema`.

let me know your thoughts! Thanks!

Reply:
@ThrawnCA @duttonw okay sorry, I realize that code is only used in the badge helper.

https://github.com/qld-gov-au/ckanext-validation/blob/master/ckanext/validation/logic/action.py#L53
So in the `resource_validation_run` it just returns out if no `schema`.

Just to confirm: these no-schema resources still go to XLoader and DataStore??

Reply:
Sure, if it's a tabular format then XLoader will consume it, no schema required.